drivers/nec/display/display.cr

```diff
@@ -0,0 +1,397 @@
+require "placeos-driver/interface/switchable"
+
+class Nec::Display::All < PlaceOS::Driver
+  include Interface::Powerable
+  include Interface::AudioMuteable
+
+  enum Inputs
+    Vga          = 1
+    Rgbhv        = 2
+    Dvi          = 3
+    Hdmi_set     = 4
+    Video1       = 5
+    Video2       = 6
+    Svideo       = 7
+    Tuner        = 9
+    Tv           = 10
+    Dvd1         = 12
+    Option       = 13
+    Dvd2         = 14
+    Display_port = 15
+    Hdmi         = 17
+    Hdmi2        = 18
+    Hdmi3        = 130
+    Usb          = 135
+
+    def to_s : String
+      Nec::Display::All.format_value(self.value)
+    end
+  end
+  include PlaceOS::Driver::Interface::InputSelection(Inputs)
+
+  # Discovery Information
+  tcp_port 7142
+  descriptive_name "NEC Display"
+  generic_name :Display
+
+  default_settings({
+    volume_min: 0,
+    volume_max: 100
+  })
+
+  @volume_min : Int32 = 0
+  @volume_max : Int32 = 100
+
+  @target_input : Input? = nil
+  @target_audio : Audio? = nil
+  @input_double_check : PlaceOS::Driver::Proxy::Scheduler? = nil
+
+  DELIMITER = 0x0D_u8
+
+  def on_load
+    # Communication settings
+    queue.delay = 120.milliseconds
+    # 0x0D (<CR> carriage return \r)
+    transport.tokenizer = Tokenizer.new(Bytes[DELIMITER])
+    on_update
+  end
+
+  def on_update
+    @volume_min = setting(Int32, :volume_min)
+    @volume_max = setting(Int32, :volume_max)
+  end
+
+  def connected
+    schedule.every(50.seconds, true) do
+      do_poll
+    end
+  end
+
+  def disconnected
+    schedule.clear
+  end
+
+  def power(state : Bool)
+    data = "C203D6"
+    current = self[:power]?
+
+    if current
+      data += "0004" # 0004 = Power Off
+      do_send(MsgType::Command, data, name: "power", delay: 10.seconds, timeout: 10.seconds)
+
+      self[:power] = false
+      logger.debug { "-- NEC LCD, requested to power off" }
+    else
+      data += "0001" # 0001 = Power Off
+      do_send(MsgType::Command, data, name: "power", delay: 5.seconds)
+      self[:warming] = true
+      self[:power] = true
+      logger.debug { "-- NEC LCD, requested to power on" }
+
+      power_on_delay
+      mute_status(20)
+      volume_status(20)
+    end
+  end
+
+  def power?(**options)
+    do_send(MsgType::Command, Command::Power_query, **options)
+  end
+
+  def switch_to(input : Input)
+    @target_input = input
+    @target_audio = nil
+
+    data = Command::Video_input.to_s + input.to_s
+
+    do_send(MsgType::Set_parameter, data, name: "input", delay: 6.seconds)
+    video_input
+
+    # Double check the input again!
+    # @input_double_check.cancel if @input_double_check
+    # @input_double_check = schedule.in(4.seconds) do
+    #   @input_double_check = nil
+    #   video_input
+    # end
+    # TODO: check if the below is equivalent to what's above
+    @input_double_check.try &.terminate
+    schedule.in(4.seconds) do
+      @input_double_check = nil
+      video_input.as(PlaceOS::Driver::Transport)
+    end
```

Comment:
As per the previous review - this can be dropped. If there's issue with a device not correctly responding, this should be handles as part of the retry behaviour for the input query.